File: apps/android/app/src/test/java/com/dndsync/android/pair/PairTestDoubles.kt
package com.dndsync.android.pair

class InMemoryPairStore : PairStoring {
    var stored: StoredPair? = null

    override fun load(): StoredPair? = stored
    override fun save(pair: StoredPair) {
        stored = pair
    }
    override fun clear() {
        stored = null
    }
}

class FakePairForwarder : PairForwarder {
    var joined = false
    var registered = false
    var joinCount = 0
    var registerCount = 0
    var deleted = false
    var envelopes = 0
    var error: Exception? = null

    override fun registerDevice(
        baseUrl: String,
        pairId: String,
        secret: String,
        sender: String,
        platform: String,
        token: String,
        e2ePublicKey: String?,
    ) {
        registerCount += 1
        error?.let { throw it }
        registered = true
    }

    override fun joinPair(
        baseUrl: String,
        pairId: String,
        secret: String,
        sender: String,
        platform: String,
        token: String,
        e2ePublicKey: String,
    ) {
        joinCount += 1
        error?.let { throw it }
        joined = true
    }

    override fun postEnvelope(baseUrl: String, pairId: String, secret: String, envelope: ByteArray) {
        error?.let { throw it }
        envelopes += 1
    }

    override fun deletePair(baseUrl: String, pairId: String, secret: String) {
        error?.let { throw it }
        deleted = true
    }
}

File: apps/macos/FocusSyncTests/PairTestDoubles.swift
import Foundation
@testable import FocusSync

final class InMemoryPairStore: PairStoring {
    var stored: PersistedPair?

    func load() -> PersistedPair? { stored }
    func save(_ pair: PersistedPair) { stored = pair }
    func clear() { stored = nil }
}

final class FakePairForwarder: PairForwarder {
    var devices: [ForwarderDevice] = []
    var created = false
    var deleted = false
    var envelopes: [Data] = []
    var registeredTokens: [String] = []
    var apnsEnvironments: [String] = []
    var error: Error?
    /// Applied only to `listDevices`, so a 401 while checking the peer can
    /// still be followed by a successful `createPair`.
    var listError: Error?

    func registerDevice(
        baseURL: String,
        pairId: String,
        secret: String,
        sender: String,
        platform: String,
        token: String,
        e2ePublicKey: String?,
        apnsEnvironment: String
    ) async throws {
        if let error { throw error }
        registeredTokens.append(token)
        apnsEnvironments.append(apnsEnvironment)
    }

    func createPair(
        baseURL: String,
        appKey: String,
        pairId: String,
        secretHash: String,
        sender: String,
        platform: String,
        token: String,
        e2ePublicKey: String?,
        apnsEnvironment: String
    ) async throws {
        if let error { throw error }
        created = true
        apnsEnvironments.append(apnsEnvironment)
    }

    func listDevices(
        baseURL: String,
        pairId: String,
        secret: String
    ) async throws -> [ForwarderDevice] {
        if let listError { throw listError }
        if let error { throw error }
        return devices
    }

    func postEnvelope(
        baseURL: String,
        pairId: String,
        secret: String,
        envelope: Data
    ) async throws {
        if let error { throw error }
        envelopes.append(envelope)
    }

    func deletePair(baseURL: String, pairId: String, secret: String) async throws {
        if let error { throw error }
        deleted = true
    }
}
